Support for Dropwizard Instrumentation with 1.0.0

kingster opened the pull request:
Attempt to Solve #867

@fstab Hi, this is my first time with prometheus library, so please pardon any of my mistakes. Need a little help from you in validating the approach.  I am trying to port the library to support with the new 1.0 models and I tried to migrate to `MultiCollector` as discussed in #867 , Regarding tests, I am planning to write asserts with conversion to `convertToOpenMetricsFormat` (I couldn't find any alternative way to directly access the values without converting to string).

~~`CustomMapping` I will attempt once the basic dropwizard-export works perfectly fine.~~ Added Support for this too.

fstab reviewed (changes requested):
Thanks a lot, this is wonderful!

The approach looks great, I think this is doing the right thing. I still need to do a line-by-line review, but in general it's looking good.

Two initial remarks:

* This is an instrumentation, not an exporter. Please rename the module to `prometheus-metrics-instrumentation-dropwizard`.
* I would like to keep the Git history intact, therefore I kept the old implementation in `simpleclient-archive/`. Could you start with
```
git mv simpleclient-archive/simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java prometheus-metrics-instrumentation-dropwizard/src/main/java/io/prometheus/metrics/instrumentation/dropwizard/DropwizardExports.java
```
and then apply your changes? I think this should keep the history intact. Same for the test and other files if applicable. Once everything is ported, you can remove `simpleclient-archive/simpleclient_dropwizard/`.

I'll get back to you with a more detailed review.

kingster commented:
Hi @fstab 

I have updated the module name to `prometheus-metrics-instrumentation-dropwizard`. The way I have migrated `simpleclient-archive/simpleclient_dropwizard` now is that I applied git mv first for the entire module and then applied all my changes, however the git diff at the PR level shows that files like `DropwizardExports.java` are newly added however at an individual commit level it shows the diff (I suppose that the changes completely changes the files and hence it is presenting that way, hence I have squasted the commits yet, once you review I will squash the commits)

I have implemented the customlabel mapping also, so this PR now completely replaces `simpleclient_dropwizard`. Please review.

fstab commented on prometheus-metrics-instrumentation-dropwizard/src/main/java/io/prometheus/metrics/instrumentation/dropwizard/DropwizardExports.java:
```diff
@@ -0,0 +1,180 @@
+     *
+     * @param registry a metric registry to export in prometheus.
+     */
+    public DropwizardExports(MetricRegistry registry) {
+        super();
+        this.registry = registry;
+        this.metricFilter = MetricFilter.ALL;
+        this.labelMapper = Optional.empty();
+    }
+
+    /**
+     * Creates a new DropwizardExports with a custom {@link MetricFilter}.
+     *
+     * @param registry     a metric registry to export in prometheus.
+     * @param metricFilter a custom metric filter.
+     */
+    public DropwizardExports(MetricRegistry registry, MetricFilter metricFilter) {
+        this.registry = registry;
+        this.metricFilter = metricFilter;
+        this.labelMapper = Optional.empty();;
+    }
+
+    /**
+     * @param registry    a metric registry to export in prometheus.
+     * @param labelMapper a labelMapper to use to map labels.
+     */
+    public DropwizardExports(MetricRegistry registry, CustomLabelMapper labelMapper) {
+        this.registry = registry;
+        this.metricFilter = MetricFilter.ALL;
+        this.labelMapper = Optional.ofNullable(labelMapper);
+    }
```
Thinking about it, maybe we should make the classes in the `labels` package not public (remove them from the public API and use them only internally), and make the label mapper configurable via the `DropwizardExports.Builder` instead. Something like this in `Builder`:

```java
        private static final ArrayList<MapperConfig> mapperConfigs = new ArrayList<>();

        /**
         * @param labels list of key/value pairs. Must be an even number.
         */
        public Builder mapperConfig(String match, String name, String... labels) {
            Map<String, String> labelsMap = new HashMap<>();
            for (int i=0; i*2<labels.length; i++) {
                labelsMap.put(labels[2*i], labels[2*i+1]);
            }
            mapperConfigs.add(new MapperConfig(match, name, labelsMap));
            return this;
        }
```

Then users could do something like this:

```java
        DropwizardExports.builder()
                .mapperConfig("test.dispatcher.*.*.*", "dispatcher_events_total", "action", "${1}", "outcome", "${2}_out")
                .register();
```

Just thinking out loud here, not sure if this is a good idea. But in general, the less classes are `public` the less backwards-compatible API we'll need to maintain. So what do you think?

kingster replied:
Added a `Builder` for `DropwizardExports`, this makes sense. However, if we want to make `labelMapper` via builder, it breaks the backward compatibility and ease of migration for people who are already using `client_java`. imho this could be a future release which makes it available via builder and let current users simply migrate by changing the import package names.

fstab reviewed (changes requested):
Thanks a lot! I added a few comments with thoughts, let me know what you think.

kingster commented on prometheus-metrics-instrumentation-dropwizard/pom.xml:
```diff
@@ -5,17 +5,21 @@
     </description>
 
+    <properties>
+        <automatic.module.name>io.prometheus.metrics.exporter.dw</automatic.module.name>
```
Got it, fixed this.

kingster commented on prometheus-metrics-instrumentation-dropwizard/pom.xml:
```diff
@@ -29,17 +33,25 @@
+            <groupId>io.dropwizard.metrics5</groupId>
             <artifactId>metrics-core</artifactId>
-            <version>4.2.9</version>
+            <version>5.0.0</version>
```
I don't see an easy way to make it compatible with both 4 and 5. The library versions are completely different and not even compatible with each other. Went ahead and renamed the package to `prometheus-metrics-instrumentation-dropwizard5`

kingster commented on prometheus-metrics-instrumentation-dropwizard/src/main/java/io/prometheus/metrics/instrumentation/dropwizard/DropwizardExports.java:
```diff
@@ -0,0 +1,180 @@
+    private static final Logger LOGGER = Logger.getLogger(DropwizardExports.class.getName());
+    private MetricRegistry registry;
+    private MetricFilter metricFilter;
+    private Optional<CustomLabelMapper> labelMapper;
```
Fixed, corrected this. Thanks

fstab reviewed (approved):
Ok, looking good, let's merge it. Thanks a lot!
